feat(change-wi-type): user can change the wi type.

divyanshiGupta opened the pull request:
<strong>Note: If there are pending changes to the PR, prefix the PR title with "WIP" and add the label "DO NOT MERGE"</strong>

### Mandatory
 - [x] What does this PR do?
        This PR integrates UI for change work item 
         type feature in planner quick preview and 
         detail page.

- [x] What issue/task does this PR references?
       Parent Story - https://openshift.io/openshiftio/Openshift_io/plan/detail/43
        Task - https://openshift.io/openshiftio/Openshift_io/plan/detail/65
- [x] Are the tests Included?
        Yes
___
### Optional
- [ ] Is the documentation Included?

- [ ] Are the Release Notes included?
<!-- For inclusion in marketing announcement - N/A for bugs. -->
<!-- A brief two line documentation of the functionality added/improved -->

- [ ] @mention(s) to expected reviewer(s) included

alien-ike commented:
### Ike Plugins (test-keeper)

Thank you @divyanshiGupta for this contribution!

<img align="left" src="https://raw.githubusercontent.com/arquillian/ike-prow-plugins/master/docs/images/arquillian_ui_failure_64px.png">

It appears that no tests have been added or updated in this PR.

Automated tests give us confidence in shipping reliable software. Please add some as part of this change.

If you are an admin or the reviewer of this PR and you are sure that no test is needed then you can use the command `/ok-without-tests` as a comment to make the status green.


Your plugin configuration is stored in the [file](https://github.com/fabric8-ui/fabric8-planner/blob/0e42af34b44dbe4933b55b5714b97c29b67f78ca/.ike-prow/test-keeper.yaml).

sanbornsen commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.ts:
```diff
@@ -78,6 +78,8 @@ export class WorkItemDetailComponent implements OnInit, OnDestroy, AfterViewChec
       this.listenToEsc = true;
     }
   }
+  @Input() workItemTypes: WorkItemTypeUI[] = [];
```
Why taking input for work item types? We have `workItemTypeSource` that gives us an observable of the list of work item types. Besides, this component will also be used for the non-contextual detail page, there, where are you gonna get the input from?

sanbornsen replied:
Or, you can use `this.wiTypes` instead.

divyanshiGupta replied:
I didnt use `workItemTypeSource` or `this.wiTypes` because they return all the work item types and not work item types for particular type group and since the code for getting work item types for a particular type group has been already implemented in the planner list component so instead of adding it in work item detail component I thought of passing it as input to work item detail component. 

But since we have to add it in the non-contextual detail page I will add the code in work item detail component as well

sanbornsen commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.ts:
```diff
@@ -78,6 +78,8 @@ export class WorkItemDetailComponent implements OnInit, OnDestroy, AfterViewChec
     }
   }
+  @Input() workItemTypes: WorkItemTypeUI[] = [];
+  selectedType: WorkItemTypeUI = null;
```
What is the use of this variable? You can directly use the relationship from work item, can't you?

divyanshiGupta replied:
I am using this variable so that as soon as the user clicks on a witype it gets updated in the UI, here
`<span class="dropdown-text">
      {{ selectedType ? selectedType.name : '' }}         
</span>
`

sanbornsen replied:
You can still do that I believe.

sanbornsen commented on src/app/effects/work-item.effects.ts:
```diff
@@ -242,13 +242,15 @@ export class WorkItemEffects {
         // We don't update work item type
         // So we remove it from the payload
-        const payload = cleanObject({
+        const payload = {
```
Remove the comments above.

sanbornsen replied:
Besides cleaning the properties, mentioned in second arg array, `cleanObject` also clean up the properties with value null. So we'd still need that function. Just remove the second arg, i.e. `['baseType']` part from it.

divyanshiGupta replied:
Sure

sanbornsen commented on src/app/components_ngrx/work-item-preview-panel/work-item-preview-panel.component.ts:
```diff
@@ -36,6 +38,7 @@ import { WorkItemUI } from './../../models/work-item';
 })
 
 export class WorkItemPreviewPanelComponent implements OnInit {
+  @Input() workItemTypes: WorkItemTypeUI[] = [];
```
We don't need this input

divyanshiGupta replied:
Will remove it

divyanshiGupta commented on src/app/components_ngrx/planner-list/planner-list.component.html:
```diff
@@ -14,7 +14,9 @@ <h4>Please wait, we are loading your data.</h4>
     </alm-planner-layout>
-    <work-item-preview-panel #quickPreview >
+    <work-item-preview-panel
+      [workItemTypes]="quickAddWorkItemTypes"
```
@sudsen I didnt get you here, what am I removing? I am just passing `quickAddWorkItemTypes` to work-item-detail component.

sanbornsen replied:
Ohh yeah, I am sorry. I thought of something and wrote something else. You don't need that input I meant.

divyanshiGupta commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.html:
```diff
@@ -120,6 +152,37 @@
               <span class="dib">
                 <strong>#{{ workItem.number }}</strong>
               </span>
+              <!-- <span class="dropdown" dropdown
```
I commented few parts of the code so that I can reuse it if need be. I will remove it in my final commit.

Raunak1203 commented:
[test]

divyanshiGupta commented:
@Raunak1203 it doesn't point to a backend story, it is the entire story consisting of frontend and backend tasks though I have updated the task reference for frontend. This is the invision link - https://redhat.invisionapp.com/share/27MZUP44AK5#/screens/309101361 for the design. I have implemented the first design @Raunak1203 @sunilmalagi

sanbornsen commented on src/tests/specs/detailPage.spec.ts:
```diff
@@ -110,4 +110,18 @@ describe('Detail View test: ', () => {
     expect(planner.detailPage.stateToggle.getTextWhenReady()).toContain('open');
   });
+
+  xit('Should change the type of work item', async () => {
```
Why is it under `xit` ?

divyanshiGupta replied:
Because since the backend is not ready the test will fail as the type will not get updated

sahil143 commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.html:
```diff
@@ -155,6 +66,96 @@
           <div class="form-group">
+            <!-- work-item-type dropdown -->
+            <section class="f8-detail__area">
+              <label class="pull-left padding-left-15 control-label">
```
This is not the correct way. Please use `common-selector` dropdown here

sahil143 commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.html:
```diff
@@ -155,6 +66,96 @@
+                    id="wi-type-dropdown"
+                    role="menu"
+                    *dropdownMenu>
+                    <li *ngFor="let type of wiTypes"
```
While using `common-selector`
wiTypes should come directly from `workItemTypeSource` i.e. `workItemTypesource | async`

sahil143 commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.html:
```diff
@@ -155,6 +66,96 @@
+          <div class="form-group">
+            <!-- work-item state dropdown -->
+            <section class="f8-detail__area">
+              <label class="pull-left padding-left-15 control-label">
```
Same as above. Use `common-selector`.

sahil143 commented on src/app/components_ngrx/work-item-detail/work-item-detail.component.ts:
```diff
@@ -364,6 +368,21 @@ export class WorkItemDetailComponent implements OnInit, OnDestroy, AfterViewChec
 
+  updateType(e: any, type: WorkItemTypeUI) {
+    if (e) {
+      e.preventDefault();
```
You don't need this in `common-selector`

sahil143 commented on src/app/effects/work-item.effects.ts:
```diff
@@ -242,15 +242,13 @@ export class WorkItemEffects {
                  ...dynamicPayload.attributes
             }}
-          }, ['baseType']);
+          });
```
If you remove `'baseType'` then on dynamic field patch request `baseType` will be included in the payload. Please check.

sanbornsen replied:
And if it stays then it won't work for type update. I don't see a problem in sending the baseType for dynamic field update.

sahil143 replied:
I'm not sure but I think It will generate an extra event for workItemType whenever a dynamic field update happens.

sanbornsen replied:
It'll not create an event for the same type, confirmed by @jarifibrahim

sahil143 commented:
@Raunak1203 I have made the requested changes.

divyanshiGupta commented:
@sahil143 what changes?

sanbornsen commented:
Closing this PR hence we got a successor - 
https://github.com/fabric8-ui/fabric8-planner/pull/2761
